### ISSUE TITLE
Overhaul logging

### ISSUE THREAD

**Issue by ttuegel** [kframework/kore#884]
An overhauled logging interface will improve performance and enable greater flexibility:

```.hs
-- | Log entries
class Typeable entry => Entry entry where
    toEntry :: entry -> SomeEntry
    toEntry = SomeEntry

    fromEntry :: SomeEntry -> Maybe entry
    fromEntry (SomeEntry entry) = Data.Typeable.cast entry

    severity :: entry -> Severity

    -- | Display the log entry nicely for the user
    displayEntry :: entry -> Doc ann

    -- | Is the log entry in the named scope?
    inScope :: String -> entry -> Bool  -- derive generically, based on constructor names

-- | Encapsulates any log entry.
data SomeEntry where
    SomeEntry :: Entry entry => entry -> SomeEntry

-- | Warning emitted when a hook is missing.
-- Using distinct types for each log entry allows the LogAction 
-- to choose how it dispatches entries. For example, the LogAction 
-- may choose not to emit the same missing-hook warning multiple times.
-- Storing data instead of text lets us push rendering into a separate thread
-- when logging is enabled. The data is also stored lazily for this reason.
-- Storing data instead of text also separates concerns between the code
-- that generates the warning and the code that displays it. Besides cleanliness,
-- this would let us change log formats quickly in the future.
data WarnMissingHook =
    WarnMissingHook { hook :: Text, symbol :: Symbol }
    deriving (Eq, Typeable)

-- | Emit the missing-hook warning.
-- See the MonadLog interface below.
warnMissingHook :: MonadLog m => Text -> Symbol -> m ()

instance Entry WarnMissingHook where
    ...

-- | Wrapper for entries in the scope of function evaluation.
data ScopeFunctionEvaluation where
    ScopeFunctionEvaluation :: Entry entry => entry -> ScopeFunctionEvaluation
    deriving (Typeable)

-- | Interface for monads that support logging.
class MonadLog m where
    log :: Entry entry => entry -> m ()
    scope :: (Entry e1, Entry e2) => (e1 -> e2) -> (m a -> m a)

-- | Implementation of a logging monad transformer.
newtype LoggerT m a =
    LoggerT { getLoggerT :: ReaderT (LogAction m SomeEntry) m a }

instance Monad m => MonadLog (LoggerT m) where
    ...
```

This interface follows the standard `Exception` interface.

Besides this change, we should push log-writing into a different thread.

See also: #872 

---

- [x] Interface
- [ ] Threading

**Comment by vladciobanu** [comment]
#980 should be the "Interface" part, so the next steps are:
- add at least one example (`WarnMissingHook`, note that we also need to make it so we only warn once per symbol)
- log on a separate thread

**Comment by ttuegel** [comment]
We should pick this up again ASAP because it has significant observability benefits. For example, right now even warnings are silenced by default.

**Comment by ttuegel** [comment]
Fixed in #1187.
